sway-core/src/semantic_analysis/namespace/trait_map.rs

```diff
@@ -5,92 +5,67 @@ use crate::{
 };
 
 type TraitName = CallPath;
-
-// This cannot be a HashMap because of how TypeInfo's are handled.
-//
-// In Rust, in general, a custom type should uphold the invariant
-// that PartialEq and Hash produce consistent results. i.e. for
-// two objects, their hash value is equal if and only if they are
-// equal under the PartialEq trait.
-//
-// For TypeInfo, this means that if you have:
-//
-// ```ignore
-// 1: u64
-// 2: u64
-// 3: Ref(1)
-// 4: Ref(2)
-// ```
-//
-// 1, 2, 3, 4 are equal under PartialEq and their hashes are the same
-// value.
-//
-// However, we need this structure to be able to maintain the
-// difference between 3 and 4, as in practice, 1 and 2 might not yet
-// be resolved.
-type TraitMapInner = im::Vector<((TraitName, TypeId), TraitMethods)>;
+/// Map of function name to [TyFunctionDeclaration](ty::TyFunctionDeclaration)
 type TraitMethods = im::HashMap<String, ty::TyFunctionDeclaration>;
+/// Map of trait name and type to [TraitMethods].
+type TraitImpls = im::HashMap<(TraitName, TypeId), TraitMethods>;
 
 #[derive(Clone, Debug, Default, PartialEq)]
 pub(crate) struct TraitMap {
-    trait_map: TraitMapInner,
+    trait_impls: TraitImpls,
 }
 
 impl TraitMap {
     pub(crate) fn insert(
         &mut self,
         trait_name: TraitName,
-        incoming_type_id: TypeId,
+        type_id: TypeId,
         methods: Vec<ty::TyFunctionDeclaration>,
     ) {
-        let mut methods_map = im::HashMap::new();
-        for method in methods.into_iter() {
-            methods_map.insert(method.name.as_str().to_string(), method);
-        }
-        self.trait_map
-            .push_back(((trait_name, incoming_type_id), methods_map));
+        let trait_methods: TraitMethods = methods
+            .into_iter()
+            .map(|method| (method.name.as_str().to_string(), method))
+            .collect();
+        let trait_impls: TraitImpls = vec![((trait_name, type_id), trait_methods)]
+            .into_iter()
+            .collect();
```

Comment:
This is totally redundant?  Did it used to have a `.map()` in there or something?  I'm surprised Clippy isn't mentioning something, unless I'm wrong.

Reply:
Looking at `im::HashMap`, it also has a `singleton` method, so even more concise would be:
```rust
let trait_impls = TraitImpls::singleton((trait_name, type_id), trait_methods);
```